libs/shared/form-fields/src/lib/CompanySearchController/CompanySearchController.tsx

```diff
@@ -0,0 +1,171 @@
+import React, { FC, useMemo, useState } from 'react'
+import { Controller, useFormContext } from 'react-hook-form'
+import {
+  AsyncSearch,
+  AsyncSearchOption,
+  Box,
+  AlertMessage,
+  InputError,
+} from '@island.is/island-ui/core'
+import { useLocale } from '@island.is/localization'
+import { coreErrorMessages } from '@island.is/application/core'
+import { gql, useLazyQuery } from '@apollo/client'
+import debounce from 'lodash/debounce'
+import kennitala from 'kennitala'
+import { CompanySearchItem } from './CompanySearchItem'
+import { debounceTime } from '@island.is/shared/constants'
+
+export const COMPANY_REGISTRY_COMPANIES = gql`
+  query SearchCompanies($input: RskCompanyInfoSearchInput!) {
+    companyRegistryCompanies(input: $input) {
+      data {
+        name
+        nationalId
+      }
+    }
+  }
+`
+
+interface Props {
+  id: string
+  defaultValue?: unknown
+  name?: string
+  label: string
+  placeholder?: string
+  initialInputValue?: string
+  inputValue?: string
+  colored?: boolean
+  setLabelToDataSchema?: boolean
+}
+
+export const CompanySearchController: FC<Props> = ({
+  defaultValue,
+  id,
+  name = id,
+  label,
+  placeholder,
+  initialInputValue = '',
+  inputValue = '',
+  colored = true,
+  setLabelToDataSchema = true,
+}) => {
+  const { clearErrors, setValue } = useFormContext()
+  const { formatMessage } = useLocale()
+
+  const [searchQuery, setSearchQuery] = useState('')
+
+  const [search, { loading, data }] = useLazyQuery(COMPANY_REGISTRY_COMPANIES)
+
+  const debouncer = useMemo(() => {
+    return debounce(search, debounceTime.search)
+  }, [search])
+
+  // Validations
+  const noResultsFound =
+    data?.companyRegistryCompanies?.data?.length === 0 &&
+    !loading &&
+    searchQuery.trim().length > 0
+  const invalidNationalId =
+    !kennitala.isValid(searchQuery) && !loading && searchQuery.length === 10
+
+  const onInputChange = (inputValue: string) => {
+    setSearchQuery(inputValue)
+    debouncer({
+      variables: {
+        input: {
+          searchTerm: inputValue,
+          first: 40,
+        },
+      },
+    })
+  }
+
+  const getSearchOptions = (
+    query: string,
+    response: { data: { name: string; nationalId: string }[] } = { data: [] },
+  ) => {
+    const { data } = response
+    const options = data?.map(({ name, nationalId }) => ({
+      label: name,
+      value: nationalId,
+      component: (props: any) => (
+        <CompanySearchItem
+          {...props}
+          key={`${name}-${nationalId}`}
+          name={name}
+          nationalId={nationalId}
+          query={query}
+        />
```

Comment:
Is AsyncSearch doing the heavy lifting for making this accessible? Including keyboard support and good screen reader UX?

Reply:
Yes, I tested it with VoiceOver and it is doing the heavy lifting :D